### PR TITLE
[WIP] Moved PAM session creations to a preexec_fn

### PR DESCRIPTION
**Pull request by cmaves**
PAM sessions are intended to be created within the spawned child
process, not in the parent process. This can create issues with
certain PAM session modules (i.e. pam-systemd.so)

### CODE REVIEW

File: jupyterhub/auth.py [hunk 1]
```diff
@@ -746,19 +749,19 @@ def authenticate(self, handler, data):
...
-    @run_on_executor
-    def pre_spawn_start(self, user, spawner):
+    def pre_exec_fn(self, user, spawner):
         """Open PAM session for user if so configured"""
         if not self.open_sessions:
             return
         try:
             pamela.open_session(user.name, service=self.service, encoding=self.encoding)
         except pamela.PAMError as e:
             self.log.warning("Failed to open PAM session for %s: %s", user.name, e)
```
Comment (minrk) [inline jupyterhub/auth.py]: Since this is now called in the subprocess, I'm not sure `self.log.warning` is going to work. Probably switch this to a lower-level print to stderr.

File: jupyterhub/auth.py [hunk 2]
```diff
@@ -356,6 +356,9 @@ def is_admin(self, handler, authentication):
                 persisted; and `admin`, the admin setting value for the user.
         """
 
+    def pre_exec_fn(self, user, spawner):
+        """Hook called before exec is called in the spawned child process """
```
Comment (minrk) [inline jupyterhub/auth.py]: ...when using a LocalProcessSpawner and its derivatives.

btw, this should probably return a pure callable rather than being an instance method calling the method. I wouldn't trust e.g. accessing the database-connection-having user object in the pre_exec_fn.

Instead, let's have something like:

```python
@staticmethod
def _preexec_fn(name, service, encoding):
    ...

def make_preexec_fn(self, user, spawner):
    return partial(self._preexec_fn, user.name, self.service, self.encoding)
```

File: jupyterhub/auth.py [hunk 3]
```diff
@@ -746,19 +749,19 @@ def authenticate(self, handler, data):
...
         except pamela.PAMError as e:
             self.log.warning("Failed to open PAM session for %s: %s", user.name, e)
-            self.log.warning("Disabling PAM sessions from now on.")
-            self.open_sessions = False
+            # this will be ineffective in the child process
+            # self.log.warning("Disabling PAM sessions from now on.")
+            # self.open_sessions = False
 
-    @run_on_executor
+    #@run_on_executor
```
Comment (minrk) [inline jupyterhub/auth.py]: This needs to come back, btw. It's important that all calls to pam happen off the main thread.

File: jupyterhub/auth.py [hunk 4]
```diff
@@ -746,19 +749,19 @@ def authenticate(self, handler, data):
...
     def post_spawn_stop(self, user, spawner):
```
Comment (minrk) [inline jupyterhub/auth.py]: As I understand it, to fix the issue, the close session also needs to happen in the subprocess as well, not just the open. How do we accomplish that? There isn't an equivalent to preexec_fn for the end of a process.

Do we instead need to create a two-layer process, where the parent does:

```python
pamela.open_session()
p = start_real_subprocess(args, env)
wait(p)
pamela.close_session()
```

File: jupyterhub/auth.py (hunk 3 above)
Comment (cmaves) [reply jupyterhub/auth.py]: The reason I removed it was because it caused this check failed with: `jupyterhub/tests/test_internal_ssl_api.py::test_bad_spawn`.

File: jupyterhub/auth.py (hunk 4 above)
Comment (cmaves) [reply jupyterhub/auth.py]: This is believe this is how `sshd` accomplishes it. It has process that watches the shell process.

I think creating a `PAMAuthenticator.post_exec_stop()` function and moving the PAM session logic into it would be best and making the two-layer process look like this:
```
if self.auth:
   self.auth.pre_exec_fn(self.user, self)
Popen.run(args, env) # will wait for process to end on its own
if self.auth:
   self.auth.post_exec_stop(self.user, self)
```

This also has the advantage of allowing me to avoid the `@run_on_executor` statement that has caused issue with one check, because the PAM session will already be occurring off of the main thread.

Comment (cmaves) [reply jupyterhub/auth.py]: It also avoid doing a lot with `pre_execfn` kwarg which the Python docs say is bad.

File: jupyterhub/auth.py (hunk 3 above)
Comment (minrk) [reply jupyterhub/auth.py]: Then that would be an issue that needs to be fixed in the test. This method must be allowed to be async. Making it a blocking method isn't an okay fix for the problem, it just ensures the async behavior isn't tested.

But this is hugely useful info! I've been wondering what the source of these errors has been, and I can look into a fix that doesn't assume an async method is synchronous, if you like.

Comment (cmaves) [reply jupyterhub/auth.py]: The most recent commit restructures this into a function that runs on a intermediate process. This make the run_on_executor statement unnecessary because it is already running of the main thread.

File: jupyterhub/auth.py (hunk 2 above)
Comment (cmaves) [reply jupyterhub/auth.py]: The instance method no longer calls this method, so this should be resolved.

File: jupyterhub/auth.py (hunk 1 above)
Comment (cmaves) [reply jupyterhub/auth.py]: I've confirmed it does work in a subprocess.

File: jupyterhub/spawner.py [hunk 5]
```diff
@@ -1308,12 +1358,17 @@ def get_env(self):
         we return the exit code of the process if we have access to it, or 0 otherwise.
         """
         # if we started the process, poll with Popen
-        if self.proc is not None:
-            status = self.proc.poll()
-            if status is not None:
-                # clear state if the process is done
+        if isinstance(self.proc_state, Queue_class):
```
Comment (minrk) [inline jupyterhub/spawner.py]: This isinstance check appears irrelevant, since proc_state is always either None or this Queue, right?

File: jupyterhub/spawner.py [hunk 6]
```diff
@@ -1308,12 +1358,17 @@ def get_env(self):
...
+            try:
+                status = self.proc_state.get(False)
+                print('PROCESS HAS FINISHED (POLL) with code', status, file=sys.stderr)
```
Comment (minrk) [inline jupyterhub/spawner.py]: debug statement

File: jupyterhub/spawner.py [hunk 7]
```diff
@@ -1308,12 +1358,17 @@ def get_env(self):
...
+                self.proc_state = status
                 self.clear_state()
-            return status
+                return status
+            except queue.Empty:
+                return None
+        elif self.proc_state is not None:
+            return self.proc_state
```
Comment (minrk) [inline jupyterhub/spawner.py]: This branch can never occur, I think

File: jupyterhub/spawner.py [hunk 8]
```diff
@@ -1254,6 +1270,36 @@ def get_env(self):
 
         return {"keyfile": key, "certfile": cert, "cafile": ca}
 
+    # intended to be run as intermediate process that
+    # calls self.authenticator.pre_exec_fn/post_exec_fn()
+    # after before and after starting the child process
+    # needed by PAMAuthenticator to handle PAM_sessions
+    def start_monitor_process(self, q, cmd, popen_kwargs):
+        if self.authenticator:
+                self.authenticator.pre_exec_fn(self.user, self)
```
Comment (minrk) [inline jupyterhub/spawner.py]: Does this pre_exec_fn need to run as the user (i.e. after setuid)?

Comment (cmaves) [reply jupyterhub/spawner.py]: Nope. PAM session modules may need elevated privileges to run such mounting drives.

**Comment by consideRatio** [comment]
Hi @cmaves and @minrk, I'm looking through PRs and hoping to nudge them towards a merge, closure, or other concrete action. Any input on this?

**Comment by cmaves** [comment]
From my point of view this pull request is dead. The core issue (#2120) still exists, but the required changes were more expansive and requires changes to how processes are spawned and forked. This should probably be done by somebody with more involvement then myself.
